### PR TITLE
Blockformatting

### PR DESCRIPTION
**Pull request by skogsmaskin**
* Format text with special text styles (header etc..) easy extendable for new styles.
* Format text as lists (easy extendable too)
* UX showing current and available formatting options

This PR should preferrably be broken into more detailed commits, but there was so much going on that I prioritised backing it up on Github with iterative commits instead. I can however break it down better with more detailed commit messages in a new branch if that is preferable.

### CODE REVIEW

File: src/inputs/BlockEditor-slate/plugins/ListItemOnEnterKey.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    return null
+  }
+  const {document, startKey, startBlock} = state
+  const previous = document.getPreviousBlock(startKey)
+  const next = document.getNextBlock(startKey)
+  if ((next && next.type === SLATE_LIST_ITEM_TYPE)
+      || !previous
+      || previous.type !== SLATE_LIST_ITEM_TYPE
+      || startBlock.text !== '') {
+    return null
```
Comment (bjoerge) [inline src/inputs/BlockEditor-slate/plugins/ListItemOnEnterKey.js]: The intention of this expression could be a bit more clear. Maybe add a comment assign each operand to descriptive variables?

File: src/inputs/BlockEditor-slate/preview/List.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import React, {PropTypes} from 'react'
+import styles from '../styles/List.css'
+
+function List(props) {
+  const listStyle = props.listStyle || props.children[0].props.parent.data.get('listStyle')
```
Comment (bjoerge) [inline src/inputs/BlockEditor-slate/preview/List.js]: Maybe add a wrapper component around the slate preview node that reads `props.children[0].props.parent.data.get('listStyle')` and passes it to this instead?

Comment (bjoerge) [inline src/inputs/BlockEditor-slate/preview/List.js]: And: add a comment describing _why_ its needed :)

File: src/inputs/BlockEditor-slate/preview/List.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  if (listStyle === 'bullet') {
+    classNames.push(styles.bullet)
+    return <ul className={classNames.join(' ')}>{props.children}</ul>
+  }
+  if (listStyle === 'roman') {
+    classNames.push(styles.roman)
+    return <ol className={classNames.join(' ')}>{props.children}</ol>
+  }
+  classNames.push(styles.number)
+  return <ol className={classNames.join(' ')}>{props.children}</ol>
```
Comment (bjoerge) [inline src/inputs/BlockEditor-slate/preview/List.js]: I'd prefer it to fail hard if I have provided an invalid listStyle.

File: src/inputs/BlockEditor-slate/preview/List.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+List.propTypes = {
+  listStyle: PropTypes.string,
```
Comment (bjoerge) [inline src/inputs/BlockEditor-slate/preview/List.js]: `PropTypes.oneOf(['bullet', 'number', 'roman'])`

File: src/inputs/BlockEditor-slate/preview/List.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  preview: PropTypes.bool,
```
Comment (bjoerge) [inline src/inputs/BlockEditor-slate/preview/List.js]: `isPreview` (or better, do this in a wrapper)

File: src/inputs/BlockEditor-slate/preview/ListItem.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+function ListItem(props) {
+  if (props.preview) {
+    return <div className={styles.root}>{props.children}</div>
+  }
+  return <li className={styles.root}>{props.children}</li>
+}
+
+ListItem.propTypes = {
+  preview: PropTypes.bool,
```
Comment (bjoerge) [inline src/inputs/BlockEditor-slate/preview/ListItem.js]: `isPreview` is clearer I think. Also, not excactly clear to the reader what it means either and why it should be a div then.

File: src/inputs/BlockEditor-slate/util/prepareSlateSchema.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
+import React from 'react'
```
Comment (skogsmaskin) [inline src/inputs/BlockEditor-slate/util/prepareSlateSchema.js]: Must be there now! :)

File: src/inputs/BlockEditor-slate/styles/ListItem.css [hunk 8]
```diff
@@ -0,0 +1,5 @@
+@use postcss-cssnext;
```
Comment (skogsmaskin) [inline src/inputs/BlockEditor-slate/styles/ListItem.css]: Not empty anymore.

**Comment by skogsmaskin** [comment]
Thank you for you review. Many nice points which are now adressed. 👍

**Pull request merged**